### ISSUE TITLE
Bundler doesn't get included?

### ISSUE THREAD

**Issue by rrauenza** [rvm/rvm1-ansible#35]
The rvm1 role attempts to create the symlinks for bundle and bundler into /usr/local/bin, but it looks like they (and the gem) haven't been installed yet:

```
TASK: [rvm1 | Symlink ruby related binaries on the system path] *************** 
failed: [prme-elab2-dhcp65] => (item=bundle) => {"failed": true, "item": "bundle", "path": "/usr/local/bin/bundle", "src": "/opt/rvm/wrappers/default/bundle", "state": "absent"}
msg: src file does not exist, use "force=yes" if you really want to create the link: /opt/rvm/wrappers/default/bundle
failed: [prme-elab2-dhcp65] => (item=bundler) => {"failed": true, "item": "bundler", "path": "/usr/local/bin/bundler", "src": "/opt/rvm/wrappers/default/bundler", "state": "absent"}
msg: src file does not exist, use "force=yes" if you really want to create the link: /opt/rvm/wrappers/default/bundler
ok: [prme-elab2-dhcp65] => (item=erb) => {"changed": false, "dest": "/usr/local/bin/erb", "gid": 0, "group": "root", "item": "erb", "mode": "0777", "owner": "root", "size": 29, "src": "/opt/rvm/wrappers/default/erb", "state": "link", "uid": 0}
```
My rubies list just contains ruby-1.9.3-p547 and rvm version is 1.26.11.  I'm using ansible 1.8.4 against a CentOS 7 box.

I'm confused -- I'm not sure how bundler is supposed to have been installed -- but it is explicitly listed in the list of scripts to symlink in the role's var file.

**Comment by nickjj** [comment]
I only started using Ruby after 2.0, is it possible that bundler is not included with 1.9.3?

**Comment by rrauenza** [comment]
I think that's the case ... Adding --with-gems=bundler to the rvm installer arguments seems to work.

I think when I was iterating the deployment at some point I must have manually installed bundler.

**Comment by rrauenza** [comment]
Was gonna close, but wasn't sure if you want to detect < 2.0 and add the option yourself since your symlink loop depends on it.

**Comment by nickjj** [comment]
I'm not a fan of solely checking `rvm1_rubies` and using it as a source of truth of the system. It gets tricky too because you can install more than 1 version of ruby.

I think we have 2 options:

- Set  `--with-gems=bundler` by default, is it smart enough to avoid extra work if ruby >= 2?
- Add documentation to warn people who are using 1.9.3 that they need to add that flag

Thoughts?

**Comment by rrauenza** [comment]
Yeah, and if you add a 1.x ruby later you'd have to reinstall rvm to actually add the --with-gems.  I feel like the rvm installer is the wrong place to be specifying this in general.

I'm not sure what happens when I set --with-gems -- it starts out as an export environment variable:

```
      (--without-gems=*|--with-gems=*|--with-default-gems=*)
        flags+=( "$token" )
        value="${token#*=}"
        token="${token%%=*}"
        token="${token#--}"
        token="${token//-/_}"
        export "rvm_${token}"="${value}"
        printf "%b" "Installing RVM ${token/_/ }: ${value}.\n"
        ;;
```
Ah -- it looks like it may go in ~/.rvmrc

```
# more ~/.rvmrc 
rvm_with_gems="bundler"
```

I just installed 2.2.1 on top of the existing rvm install and it worked fine with that .rvmrc -- I can also do this:

```
# /opt/rvm/wrappers/ruby-2.2.1/gem install bundler
Fetching: bundler-1.9.2.gem (100%)
Successfully installed bundler-1.9.2
Parsing documentation for bundler-1.9.2
Installing ri documentation for bundler-1.9.2
Done installing documentation for bundler after 7 seconds
1 gem installed
```

That could be run in a command before you do the symlinks. 

So... you could 

(a) add it as a default
(b) explicitly install the gem ( /opt/rvm/wrappers/ruby-2.2.1/gem install bundler) for every rvm1_rubies (or just the ones that are 1.x) 
(c) 'gem install' conditionally before the symlink iff the {{path}}/wrappers/{{ruby_version}}/bundle|r file doesn't exist.
(d) just document it (maybe also in the default config.)

I'd probably do (c) and maybe add a default variable for the bundler version number.  And maybe an option to turn the autobundler installer off.

I could maybe mock something up tomorrow if you like as well.

**Comment by rrauenza** [comment]
Something like this:

```
- name: Install bundler if missing
  command: '{{ rvm1_install_path }}/wrappers/default/gem' install bundler
  creates: '{{ rvm1_install_path }}/wrappers/default/bundler'
```

**Comment by rrauenza** [comment]
Maybe something like this:

```
- name: Install bundler if missing
  command: '{{ rvm1_install_path }}/wrappers/{{ item }}/gem install bundler'
  args:
    creates:
        - '{{ rvm1_install_path }}/wrappers/{{ item }}/bundler'
        - '{{ rvm1_install_path }}/wrappers/{{ item }}/bundle'
  with_items: rvm1_rubies
```

**Comment by willseward** [comment]
I'm having the same problem. I agree with @rrauenza to create the "Install bundler if missing" task.

**Comment by nickjj** [comment]
@willseward I'm guessing you're using Ruby 1.9.x as well?

@rrauenza Would you mind making a PR out of this, and also including a boolean option to the defaults? Something like `rvm1_install_bundler` , defaults to `true`. If true, your task runs. If false, it skips.

The documentation for it could be:

```
# Ruby versions prior to 2.x do not come with bundler.
# When enabled this will ensure bundler is installed.
rvm1_install_bundler: true
```

**Comment by willseward** [comment]
I'm using 1.8.7, but it's has the same effect on the role.

@nickjj Just a thought... Some applications just need the ruby environment without the bundler, like Phusion Passenger in my case. Do you think an option to exclude the bundler symlinks would be beneficial at all?

**Comment by nickjj** [comment]
I think that would complicate the options of the role a lot, there would be too many states. Right now it's a little confusing even with this new proposed task because we have 2 states:

1. Bundler gets manually installed and symlinks get created (rvm1_install_bundler: true)
2. Bundler does not get manually installed and no symlinks get created (rvm1_install_bundler: false)

But the variable name is confusing because most people are running Ruby 2.0 and may for whatever reason expect bundler to not be installed if they set it to false, but it will still be there since it's part of ruby itself.

**Comment by willseward** [comment]
I agree that is a bit confusing. 

Could this work better? Probably directly before the symlink task

```
- name: Test for bundler installation
  shell: gem list | grep "^bundler " | awk {'print $1'}
  register: bundler_result

- name: Install bundler if not installed # rrauenza's task
  command: '{{ rvm1_install_path }}/wrappers/{{ item }}/gem install bundler'
  args:
    creates:
        - '{{ rvm1_install_path }}/wrappers/{{ item }}/bundler'
        - '{{ rvm1_install_path }}/wrappers/{{ item }}/bundle'
  with_items: rvm1_rubies
  when: bundler_result == "bundler"
```

**Comment by gmalouf** [comment]
I think you mean when bundler_result != "bundler" right?

**Comment by willseward** [comment]
Yes, thanks!

**Comment by gmalouf** [comment]
I just tested this, it fails as follows:

TASK: [rvm_io.rvm1-ruby | Install bundler if not installed] ******************* 
failed: [52.5.35.135] => (item=ruby-2.0.0) => {"cmd": "/usr/local/rvm/wrappers/ruby-2.0.0/gem install bundler", "failed": true, "item": "ruby-2.0.0", "rc": 2}
msg: [Errno 2] No such file or directory

The reason is that /usr/local/rvm/wrappers contains the following:

default  ruby-2.0.0-p643  ruby-2.0.0-p643@global

I'm using default locally for my code to work, but the task needs a way to know the full ruby version (not just ruby-2.0.0 in my case).

I think instead

**Comment by nickjj** [comment]
You'll probably also need to do:
`when: bundler_result.stdout != "bundler"`

**Comment by gmalouf** [comment]
@nickjj I'll try that out now.

**Comment by byped** [comment]
Could this be related to rvm itself no longer installing bundler by default?

See: https://github.com/rvm/rvm/issues/3366

**Comment by adamniedzielski** [comment]
I experienced it with Ruby 2.2.2. I think that @byped is right - that was the cause in my case.

**Comment by nickjj** [comment]
Ok.

**Comment by udondan** [comment]
I created a pull requests to fix this in #40, based on the code in this issue.

I merged it together in one task to speed things up. As well the `creates` arg requires the value to be a string. A list as posted above fails (at least in the Ansible _devel_ branch which is the upcoming 2.0 release) Checking on value though should be sufficient, I guess.

**Comment by bfallik** [comment]
I believe this issue and the associated fix will address a build failure we're seeing. I wondering when to expect the update to appear on ansible galaxy?
  https://galaxy.ansible.com/list#/roles/1087
The latest version listed is 1.3.5 from January.

**Comment by nickjj** [comment]
@bfallik Sorry about that. I merged in his PR but didn't create a release.

It's fixed, released and up on the Galaxy as `v1.3.6`.

**Comment by bfallik** [comment]
Thanks! Turns out I'm not sure this PR addresses my issue but I'm going to retry with 1.3.6 instead of selectively applying the patch.
